### ISSUE TITLE
Mixed precision : Numerical instability with instance normalization layer

### ISSUE THREAD

**Issue by RocaVincent** [tensorflow/tensorflow#47156]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.5.0
- Python version: 3.8.5
- Bazel version (if compiling from source): 4.0.0
- GCC/Compiler version (if compiling from source): 9.3.0
- CUDA/cuDNN version: 11.1/8.0.5
- GPU model and memory: Nvidia Quadro RTX 6000 ~ 23.8G

**Describe the current behavior**
I've developed a [WGAN-GP](http://arxiv.org/abs/1704.00028) model which uses some _Instance Normalization_ layers. The code I've written for _Instance Normalization_ is given above. Also I use mixed precision to speed up computations and save GPU memory. What I've observed by experimenting my model is that these normalization layers cause important numerical instability (gradient penalty goes up after hundreds of train steps) when their _dtype_ are left to mixed precision. Indeed, when I explicitly set the dtype to "float32" in the [parent constructor](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Layer#arguments), I don't have these problems.

**Describe the expected behavior**
I don't understand why the normalization layers whose code is quite simple don't work well in mixed precision mode.

**Standalone code to reproduce the issue**
There is no simple way to reproduce the issue, I just hope to have advises, suggestions or explanations to better understand this behavior.

**Other info / logs** 
```
class InstanceNormalization(keras.layers.Layer):

    def __init__(self, epsilon=1e-3, **kwargs):
        super().__init__(**kwargs)
        self.epsilon=epsilon

    def build(self, batch_input_shape):
        self.scale = self.add_weight(
            name='scale',
            shape=batch_input_shape[-1:],
            initializer=tensorflow.random_normal_initializer(1,0.02),
            trainable=True
        )
        self.offset = self.add_weight(
            name='offset',
            shape=batch_input_shape[-1:],
            initializer="zeros",
            trainable=True
        )
        self.axis = range(1, len(batch_input_shape)-1)
        super().build(batch_input_shape)

    def call(self, x):
        mean = keras.backend.mean(x, axis=self.axis, keepdims=True)
        variance = keras.backend.mean(keras.backend.square(x-mean), axis=self.axis, keepdims=True)
        normalized = (x - mean) / keras.backend.sqrt(variance+self.epsilon)
        return self.scale * normalized + self.offset
```

**Comment by amahendrakar** [comment]
@RocaVincent,
Without a reproducible code it would be difficult for us to determine the source of the issue. 

In order to expedite the trouble-shooting process, could you please provide a minimal code snippet to reproduce the issue reported here. Thanks!

**Comment by RocaVincent** [comment]
Hi @amahendrakar,
As I said in the first post, this issue is not reproducible with a minimal code. However, I can assure you that there is no training instability when I made this layers computing in float32 (this find took a long time). The results are clear and are very surprising because these layers are just a small part of my model. Therefore, I've shared the layer's code in the case you have any explanations concerning the bad results with mixed precision.

**Comment by jvishnuvardhan** [comment]
@RocaVincent Can you please check whether this was an issue (regression) with earlier version like 2.3 or 2.4? thanks!

**Comment by reedwm** [comment]
For the nonfused cases, Tensorflow does both [BatchNormalization](https://github.com/tensorflow/tensorflow/blob/c20905d3da8027bb8cd1b2de71f266a64cd7d0fe/tensorflow/python/keras/layers/normalization.py#L783-L786) and [LayerNormalization](https://github.com/tensorflow/tensorflow/blob/c20905d3da8027bb8cd1b2de71f266a64cd7d0fe/tensorflow/python/keras/layers/normalization.py#L1236-L1238) in float32 internally. Other than variance overflowing, which results in NaNs, I am unsure if this needed and if so, why it needs to be float32. But the fact InstanceNormalization is running into issues indicates there is probably some underlying issue with normalization in float16.

@RocaVincent I doubt variance is overflowing in this case since I've only observed this happen near the start of training, but if you want you can check with [`tf.debugging.check_numerics`](https://www.tensorflow.org/api_docs/python/tf/debugging/check_numerics). I suspect there is some other issue but unfortunately I don't know what it is.

@nluehr do you know why InstanceNormalization would have issues in float16, and whether BatchNormalization and LayerNormalization would run into the same issue (other than variance overflowing)? Also, when fused, do you know what parts of cudnn's BatchNormalization are done in float32 and what parts are in float16? It would be nice to root cause this so we could give advice in the mixed precision documentation, or perhaps somehow give a warning/error in these cases.

> @RocaVincent Can you please check whether this was an issue (regression) with earlier version like 2.3 or 2.4? thanks!

I don't think this is necessary, since it's time consuming and I doubt any changes to TF would have caused this issue.

**Comment by RocaVincent** [comment]
@reedwm I just tested what you mention with *tf.debugging.check_numerics* and indeed variance overflowed quite quickly. It's certainly the cause of the GAN divergence even if this divergence happens only after lots of train steps.

> Other than variance overflowing, which results in NaNs, I am unsure if this needed and if so,
Do you think mean computation (not only in InstanceNormalization) can also product overflow ?

**Comment by reedwm** [comment]
Thank you for the update! I am unsure why the variance overflows quickly but the gradient penalty takes longer to diverge. It is possible that the NaN or Inf from the overflow is removed in a future layer, making all values finite and allowing the model to partially train. I am also currently unfamiliar with WGAN-GP and gradient penalty in general, so it could be something specific to how gradient penalty is computed.

> Do you think mean computation (not only in InstanceNormalization) can also product overflow ?

The mean computation will not overflow, as the mean cannot possibly be larger in magnitude than the largest input value. However the [NVIDIA mixed precision guide](https://docs.nvidia.com/deeplearning/performance/mixed-precision-training/index.html) claims both mean and variance should be in float32. I'm not sure why, but better to do mean in float32 to be safe, I think. @nluehr do you know why the NVIDIA guide recommends computation the mean in float32?

**Comment by RocaVincent** [comment]
The tests I have done for few days show that the mean computation may overflow (wheter with `keras.backend.mean` or `tensorflow.reduce_mean`), even if there is no overflow in the input. It's certainly due to the sum before division. It's a bad thing beacause all these casts to float32 and float16 (for every mean computation) dtype make the model much more heavier.

**Comment by jvishnuvardhan** [comment]
@RocaVincent Is this still an issue for you? Can you please check with recent TF versions (`TF2.7` , `tf-nightly`) and let us know whether it is persisting with recent TF versions. Thanks!
